File: leetcode/0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.java
class Solution {
    public void flatten(TreeNode root) {
        TreeNode curr = root;

        while (curr != null) {
            if (curr.left != null) {
                // Find the rightmost node of the left subtree
                TreeNode leftRightMost = curr.left;
                while (leftRightMost.right != null) {
                    leftRightMost = leftRightMost.right;
                }

                // Connect the original right subtree
                leftRightMost.right = curr.right;

                // Move the left subtree to the right
                curr.right = curr.left;
                curr.left = null;
            }

            // Move to the next node
            curr = curr.right;
        }
    }
}

File: leetcode/0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.js
var flatten = function(root) {
    let curr = root;
    while(curr){
        if(curr.left){
            // 왼쪽 노드들의 가장 오른쪽에 현재node오른쪽 부분을 이어붙여야 하기 때문에 포인터 이동
            let l = curr.left;
            while(l.right) l = l.right;
            
            l.right = curr.right;  // 현재노드의 왼편중에서 가장 오른쪽에다가 현재오른노드를 붙여준다.
            curr.right = curr.left;  // 왼쪽노드를 오른쪽으로 이동
            curr.left = null;  // 왼쪽 비우기
        }
        curr = curr.right;  // 계속 오른쪽으로 넘어가면서 탐색. 왼쪽노드가 있을때마다 위의 루프 실행예정
    }
};